src/Messages/OWSMessageSender.m

```diff
@@ -678,14 +652,20 @@ - (void)groupSend:(NSArray<SignalRecipient *> *)recipients
     [self saveGroupMessage:message inThread:thread];
     NSMutableArray<TOCFuture *> *futures = [NSMutableArray array];
 
-    for (SignalRecipient *rec in recipients) {
+    for (SignalRecipient *recipient in recipients) {
         // We don't need to send the message to ourselves...
-        if ([rec.uniqueId isEqualToString:[TSStorageManager localNumber]]) {
+        if ([recipient.uniqueId isEqualToString:[TSStorageManager localNumber]]) {
+            continue;
+        }
+        if ([message wasSentToRecipient:recipient.uniqueId]) {
+            // Skip recipients we have already sent this message to (on an
+            // earlier retry, perhaps).
+            DDLogInfo(@"Skipping group message recipient; already sent: %@", recipient.uniqueId);
```

Comment:
...And we use that bookkeeping to reduce (but not completely avoid) double-sending.